### PR TITLE
Support classifying message toxicity locally

### PR DESCRIPTION
**Pull request by mia-pi-git**
This PR both adds a local message classifier, and moves the code for the remote classifier using Perspective to a directory in server/. The local classifier is in server/artemis/local.ts, the remote is in server/artemis/remote.ts. The local classifier uses a python dependency to classify messages - the TS file uses `child_process.spawn` to wrap the script. It is kept in a StreamProcessManager, where each stream spawns an instance of the python process. This is done so the spawning isn't done in the main process, since that causes noticeable lag. The remote code was just copy/pasted from the abuse-monitor chat plugin.

### CODE REVIEW

File: server/artemis/install [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env sh
+cd server/artemis/
```
Comment (AnnikaCodes) [inline server/artemis/install:2]:
```suggestion
#!/usr/bin/env bash
set -o errexit
set -o errtrace
cd server/artemis/
```

File: server/artemis/model.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+def now():
+	return int(time.time() * 1000)
+
+logfile = open("logs/artemis.log", "a")
+def log(message):
+	if not debug:
+		return None
+
+	logfile.write(str(now()) + ":" + str.rstrip(message) + "\n")
+	logfile.flush()
```
Comment (AnnikaCodes) [inline server/artemis/model.py:31]:
```suggestion
	if debug:
		logfile.write(str(now()) + ":" + str.rstrip(message) + "\n")
		logfile.flush()
```

File: server/artemis/model.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	for line in sys.stdin:
+		log("in:" + line)
+		parts = line.split("|")
+		out = parts.pop(0) + "|"
+		try:
+			res = model.predict("|".join(parts))
+			for key in res: res[key] = str(res[key]) # json.dumps doesn't like floats
+			out += json.dumps(res)
+		except BaseException as e:
+			out += json.dumps({'error': f"{e}"})
```
Comment (AnnikaCodes) [inline server/artemis/model.py:45]: If you're using f-strings here (thus preventing the script from running on Python version that don't support f-strings), why not use them for the rest of the string interpolation you do?

File: server/artemis/remote.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+		return PM.respawn();
+	}
+	spawn(number: number) {
+		PM.spawn(number);
+	}
+	getActiveProcesses() {
+		return PM.processes.length;
+	}
+}
+
```
Comment (AnnikaCodes) [inline server/artemis/remote.ts:176]: I assume this file is just copy/pasted from the preexisting Artemis code? Are there changes I should be aware of?

File: server/artemis/local.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+			// floats have to be made into strings because python's json.dumps
+			// doesn't like float32s
+			data[k] = parseFloat(data[k]);
+		}
+		return data as Record<string, number>;
+	}
+}
+
+// main module check necessary since this gets required in other non-parent processes sometimes
+// when that happens we do not want to take over or set up or anything
```
[the remark this answers is not in the document]
Comment (mia-pi-git) [reply server/artemis/local.ts:135]: No. It's not one or the other, it's both for different applications.

File: server/artemis/remote.ts (hunk 4 above)
Comment (mia-pi-git) [reply server/artemis/remote.ts:176]: No. It's all the same.

File: server/artemis/model.py (hunk 3 above)
Comment (mia-pi-git) [reply server/artemis/model.py:43]: Yeah, probably should. Done in https://github.com/smogon/pokemon-showdown/pull/8662/commits/ae3f15d7577854f769423d08b4c5220a64cf4b74.

File: server/artemis/local.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (AnnikaCodes) [reply server/artemis/local.ts:142]: Oh, okay.

File: server/artemis/local.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+				const info = JSON.parse(dataStr);
+				Monitor.crashlog(new Error(info.error), "An Artemis script", info);
+				try {
+					this.pushEnd(); // push end first so the stream always closes
+					this.process.disconnect();
+				} catch {}
+			}
+		});
+		this.process.stderr.on('data', data => {
+			data += "";
```
Comment (AnnikaCodes) [inline server/artemis/local.ts:43]: ?

Comment (mia-pi-git) [reply server/artemis/local.ts:43]: data is a Buffer by default. This just makes it into a string.

Comment (AnnikaCodes) [reply server/artemis/local.ts:43]: Why choose `+= ""` instead of `.toString()`?

Comment (mia-pi-git) [reply server/artemis/local.ts:43]: Looks nicer, in my opinion.

**Review by AnnikaCodes — APPROVED** [review]
This looks fine to me, assuming it's been tested.

File: server/artemis/local.ts (hunk 6 above)
Comment (AnnikaCodes) [reply server/artemis/local.ts:43]: I think it makes the code more difficult to read for people who aren't familiar with JavaScript's type coercion.

Comment (mia-pi-git) [reply server/artemis/local.ts:43]: We can junk it anyway since setEncoding('utf-8') means the data is a string. This was just a prior relic of that.

**Comment by mia-pi-git** [comment]
> This looks fine to me, assuming it's been tested.

Yes, I've done a bunch of testing.

**Comment by AnnikaCodes** [comment]
Thanks!

**Pull request merged**